x-pack/plugins/security_solution/public/common/components/exceptions/viewer/index.tsx

```diff
@@ -4,96 +4,411 @@
-  handleEdit: (item: ExceptionListItemSchema) => void;
+  onAssociateList?: (listId: string) => void;
 }
 
-const ExceptionItemComponent = ({
-  exceptionItem,
+const ExceptionsViewerComponent = ({
+  ruleId,
+  exceptionListsMeta,
+  availableListTypes,
+  onAssociateList,
   commentsAccordionId,
-  handleDelete,
-  handleEdit,
-}: ExceptionItemProps): JSX.Element => {
-  const [entryItems, setEntryItems] = useState<FormattedEntry[]>([]);
-  const [showComments, setShowComments] = useState(false);
+}: ExceptionsViewerProps): JSX.Element => {
+  const { services } = useKibana();
+  const [, dispatchToaster] = useStateToaster();
+  const [initLoading, setInitLoading] = useState(true);
+  const onDispatchToaster = useCallback(
+    ({ title, color, iconType }) => (): void => {
+      dispatchToaster({
+        type: 'addToaster',
+        toast: {
+          id: uuid.v4(),
+          title,
+          color,
+          iconType,
+        },
+      });
+    },
+    [dispatchToaster]
+  );
+  const { deleteExceptionItem } = useApi(services.http);
+  const [
+    {
+      endpointList,
+      detectionsList,
+      exceptions,
+      filterOptions,
+      pagination,
+      loadingItemIds,
+      isModalOpen,
+    },
+    dispatch,
+  ] = useReducer(allExceptionItemsReducer(), initialState);
 
-  useEffect((): void => {
-    const formattedEntries = getFormattedEntries(exceptionItem.entries);
-    setEntryItems(formattedEntries);
-  }, [exceptionItem.entries]);
+  const setExceptions = useCallback(
+    ({
+      lists: newLists,
+      exceptions: newExceptions,
+      pagination: newPagination,
+    }: SetExceptionsProps) => {
+      dispatch({
+        type: 'setExceptions',
+        lists: newLists,
+        exceptions: newExceptions,
+        pagination: newPagination,
+      });
+    },
+    [dispatch]
+  );
+  const [loadingList, , , , fetchList] = useExceptionList({
+    http: services.http,
+    lists: exceptionListsMeta,
+    filterOptions,
+    pagination: {
+      page: pagination.pageIndex + 1,
+      perPage: pagination.pageSize,
+      total: pagination.totalItemCount,
+    },
+    dispatchListsInReducer: setExceptions,
+    onError: onDispatchToaster({
+      color: 'danger',
+      title: i18n.FETCH_LIST_ERROR,
+      iconType: 'alert',
+    }),
+  });
 
-  const onDelete = useCallback((): void => {
-    handleDelete({ id: exceptionItem.id });
-  }, [handleDelete, exceptionItem]);
+  const setIsModalOpen = useCallback(
+    (isOpen: boolean): void => {
+      dispatch({
+        type: 'updateModalOpen',
+        isOpen,
+      });
+    },
+    [dispatch]
+  );
+
+  const onFetchList = useCallback((): void => {
+    if (fetchList != null) {
+      fetchList();
+    }
+  }, [fetchList]);
+
+  const onFiltersChange = useCallback(
+    ({ filter, pagination: pag }: Filter): void => {
+      dispatch({
+        type: 'updateFilterOptions',
+        filterOptions: filter,
+        pagination: pag,
+      });
+    },
+    [dispatch]
+  );
```

Comment:
I mentioned this in another comment in reference to updating the search filter, but we'll want to show the loader as part of this action since it results in a network call. This should cover pagination, list-type filters, and search filter updates.

Reply:
Should I make a request when a user toggles list types? I had looked at the all rules table and thought I noticed that it doesn't refresh on the filter toggles.

Reply:
To keep it simple I would just re-request on selection (i.e. any filterOptions changes). If this was an in-memory table you'd be fine, but say there's two lists linked to the rule, one `detection` with 10 items, and one `endpoint` with 2000 items, and an `items per page` of 5 you'll definitely have to.

As for the `All Rules` table, I just checked and it's making a request on any filterOption change, be it tags, filters, or pagination. I wonder if we've got a 🐛 in there 😬 ...

Reply:
Gaah, no 🐛 , I just had the Network tab filtering for `exceptions` 😬  Updated in follow up PR to fetch on all filter changes.